arangod/Aql/ClusterQuery.cpp

```diff
@@ -45,43 +46,47 @@ using namespace arangodb::aql;
 ClusterQuery::ClusterQuery(QueryId id,
                            std::shared_ptr<transaction::Context> ctx,
                            QueryOptions options)
-    : Query(id, ctx, aql::QueryString(),
-            /*bindParams*/ nullptr, std::move(options),
+    : Query{id,
+            ctx,
+            {},
+            /*bindParams*/ nullptr,
+            std::move(options),
             /*sharedState*/ ServerState::instance()->isDBServer()
                 ? nullptr
-                : std::make_shared<SharedQueryState>(ctx->vocbase().server())) {
-}
-
-ClusterQuery::~ClusterQuery() {
-  try {
-    _traversers.clear();
-  } catch (...) {
-  }
+                : std::make_shared<SharedQueryState>(ctx->vocbase().server())} {
 }
 
 /// @brief factory method for creating a cluster query. this must be used to
 /// ensure that ClusterQuery objects are always created using shared_ptrs.
-/*static*/ std::shared_ptr<ClusterQuery> ClusterQuery::create(
+std::shared_ptr<ClusterQuery> ClusterQuery::create(
     QueryId id, std::shared_ptr<transaction::Context> ctx,
-    aql::QueryOptions options) {
-  // workaround to enable make_shared on a class with a private/protected
-  // constructor
-  struct MakeSharedQuery : public ClusterQuery {
+    QueryOptions options) {
+  // workaround to enable make_shared on a class with a protected constructor
+  struct MakeSharedQuery final : ClusterQuery {
     MakeSharedQuery(QueryId id, std::shared_ptr<transaction::Context> ctx,
-                    aql::QueryOptions options)
-        : ClusterQuery(id, std::move(ctx), std::move(options)) {}
+                    QueryOptions options)
+        : ClusterQuery{id, std::move(ctx), std::move(options)} {}
+
+    ~MakeSharedQuery() final {
+      try {
+        _traversers.clear();
+      } catch (...) {
+      }
```

Comment:
I think this should be kept in the ClusterQuery dtor as before. Even though TSan reported no related races (yet), I find it rather discomforting that we destroy the traversers of a query _while it might still be accessed_, because we are only waiting for the query to finish execution in `destroy`.

Reply:
Done